articles/concepts-grovers.md

```diff
@@ -0,0 +1,185 @@
+
+$$O_f = R_{\ket{\text{bad}}} = 2\ket{\text{bad}}\bra{\text{bad}} - \mathcal{I}$$
+
+Analogously, the conditional phase shift $O_0$ is just an inverted reflection about the state $\ket{0}$:
+
+$$O_0 = R_{\ket{0}}=  -2\ket{0}\bra{0} + \mathcal{I}$$
+
+Knowing this fact, it's easy to check that the Grover's diffusion operation $-H^{\otimes n} O_0 H^{\otimes n}$ is also a reflection about the state $\ket{all}$. Just do:
+
+$$H^{\otimes n} O_0 H^{\otimes n}=2H^{\otimes n}\ket{0}\bra{0}H^{\otimes n}
+-H^{\otimes n}\mathcal{I}H^{\otimes n} = 2\ket{\text{all}}\bra{\text{all}} - \mathcal{I} = R_{\ket{\text{all}}}$$
+
+We just demonstrated that each iteration of the Grover's algorithm is a composition of two rotations $R_\ket{\text{bad}}$ and $R_\ket{\text{all}}$.
+
+![](./media/grovers-iteration.png)
+
+The resulting operation after each Grover's iteration is a counterclockwise
+rotation of an angle $2\theta$. Fortunately, the angle $\theta$ is easy to find.
+Since $\theta$ is just the angle between $\ket{\text{all}}$ and $\ket{\text{bad}}$ we can use
+the scalar product to find the angle. We know that
+$\cos{\theta}=\braket{\text{all}|\text{bad}}$, so we need to calculate $\braket{\text{all}|\text{bad}}$. From the decomposition of $\ket{\text{all}}$ in terms of $\ket{\text{bad}}$ and $\ket{\text{good}}$, we get that:
+
+$$\theta = \arccos{\left(\braket{\text{all}|\text{bad}}\right)}= \arccos{\left(\sqrt{\frac{N-M}{N}}\right)} $$
+
+The angle between the state of the register and the $\ket{\text{good}}$ state will decrease in each iteration, resulting in a higher probability of measuring a valid result. To calculate this probability we just need to calculate $|\braket{\text{good}|\text{register}}|^2$. The angle $\gamma (k)$ between $\ket{\text{good}}$ and $\ket{\text{register}}$ for each iteration $k$ is just 
+$$\gamma (k) = \frac{\pi}{2}-\theta -k2\theta = \frac{\pi}{2} -(2k + 1)  \theta $$.
+
+Therefore, the probability is just:
+
+$P(\text{success}) = \cos^2(\gamma(k)) = \sin^2\left[(2k +1)\arccos \left( \sqrt{\frac{N-M}{N}}\right)\right]$
+
+## Optimal number of iterations
+
+If the probability of success as a function of the number of iterations. To find the optimal number of iterations $N_{\text{optimal}}$, we need to find the integer that maximizes it in the first cycle of the function.
+
+![](./media/success-probability-grovers.png)
+
+We know that $\sin^2{x}$ reaches it's first maximum for $x=\frac{pi}{2}$, so we just need
+to make:
+
+$$\frac{\pi}{2}=(2k_{\text{optimal}} +1)\arccos \left( \sqrt{\frac{N-M}{N}}\right)\right$$
+
+This gives us that:
+
+$$k_{\text{optimal}} = \frac{\pi}{4\arccos\left(\sqrt{1-M/N}\right)}-1/2 = \frac{\pi}{4}\sqrt{\frac{N}{M}}-\frac{1}{2}-O\left(\sqrt\frac{M}{N}\right)$$
+
+Where in the last step we used the fact that $\arccos \sqrt{1-x} = \sqrt{x} + O(x^{3/2})$.
+
+Therefore we can pick $N_\text{optimal}$ to be $N_\text{optimal} = \lfloor\lfloor \frac{\pi}{4}\sqrt{\frac{N}{M}}-\frac{1}{2} \right\rflor$.
+
```

Comment:
I am confused. Grover's algorithm provides a quadratic speedup, with $O(\sqrt(N))$ uses of $f(x)$ to find the solution. What is $M$ here?

Reply:
As per 

> ## Geometrical explanation

To see why Grover's algorithm works let's study the algorithm from a geometrical perspective. Let $\ket{bad}$ be the superposition of all states that aren't a solution to the search problem. Supposing there are $M$ valid solutions, this is:

M is the number of valid solutions to the search problem.

Reply:
Ooh I see, I had only seen Grover's with one valid solution. All clear.